File: Grind 75 in Java/0994-rotting-oranges.java
class Solution {
    int[] dirs = new int[]{0, 1, 0, -1, 0};

    public int orangesRotting(int[][] g) {
        int m = g.length;
        int n = g[0].length;
        LinkedList<int[]> q = new LinkedList<>();
        int fresh = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (g[i][j] == 2) {
                    q.add(new int[]{i, j});
                } else if (g[i][j] == 1) {
                    fresh++;
                }
            }
        }

        int time = 0;
        while (!q.isEmpty() && fresh > 0) {
            int size = q.size();

            for (int k = 0; k < size; k++) {
                int[] cur = q.removeFirst();
                for (int d = 0; d < 4; d++) {
                    int i = cur[0] + dirs[d];
                    int j = cur[1] + dirs[d + 1];
                    if (isValid(g, i, j)) {
                        g[i][j] = 2;
                        q.addLast(new int[]{i, j});
                        fresh--;
                    }
                }                
            }

            time++;
        }

        return fresh == 0 ? time : -1;
    }

    private boolean isValid(int[][] g, int i, int j) {
        return i >= 0 && j >= 0 && i < g.length && j < g[0].length && g[i][j] == 1;
    }
}

File: leetcode in kotlin/0994-rotting-oranges.kt
class Solution {
    fun orangesRotting(grid: Array<IntArray>): Int {
        var mins = 0; var fresh = 0
        val xCord = arrayOf(1,-1,0,0)
        val yCord = arrayOf(0,0,1,-1)
        val q = ArrayDeque<Pair<Int, Int>>()
        for(r in 0..grid.size-1){
            for(c in 0..grid[0].size-1){
                if(grid[r][c] == 1)
                    fresh++
                if(grid[r][c] == 2)
                    q.add(Pair(r,c))
            }
        }
        while(!q.isEmpty() && fresh > 0){
            //we iterate all the positions that were added at time x
            repeat(q.size){
                val (x,y) = q.poll()
                for(i in 0..3){
                    val nextX = x+xCord[i]; val nextY = y+yCord[i]
                    if(isValid(grid, nextX, nextY)){
                        grid[nextX][nextY] = 2
                        q.add(Pair(nextX, nextY))
                        fresh--
                    }
                }
            }
            mins++
        }
        //did we miss some oranges that were unreachable ?
        return if(fresh == 0) mins else -1
    }
    private fun isValid(grid: Array<IntArray>, r: Int, c: Int) =
        r >= 0 && c >= 0 && r < grid.size && c < grid[0].size && grid[r][c] == 1
}
